nodefs/bridge.go

```diff
@@ -517,19 +525,120 @@ func (b *rawBridge) Fallocate(input *fuse.FallocateIn) (code fuse.Status) {
 }
 
 func (b *rawBridge) OpenDir(input *fuse.OpenIn, out *fuse.OpenOut) (status fuse.Status) {
-	return fuse.ENOSYS
+	n, _ := b.inode(input.NodeId, 0)
+	code := n.node.OpenDir(context.TODO())
+	if !code.Ok() {
+		return code
+	}
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	out.Fh = b.registerFile(nil)
+	return fuse.OK
+}
+
+func (b *rawBridge) getStream(input *fuse.ReadIn, inode *Inode, f *fileEntry) fuse.Status {
+	if f.dirStream == nil || input.Offset == 0 {
+		if f.dirStream != nil {
+			f.dirStream.Close()
+		}
+		str, code := inode.node.ReadDir(context.TODO())
+		if !code.Ok() {
+			return code
+		}
+
+		f.hasOverflow = false
+		f.dirStream = str
+	}
+
+	return fuse.OK
 }
 
 func (b *rawBridge) ReadDir(input *fuse.ReadIn, out *fuse.DirEntryList) fuse.Status {
-	return fuse.ENOSYS
+	n, f := b.inode(input.NodeId, input.Fh)
+
+	if code := b.getStream(input, n, f); !code.Ok() {
+		return code
+	}
+
+	if f.hasOverflow {
+		// always succeeds.
+		out.AddDirEntry(f.overflow)
+		f.hasOverflow = false
+	}
+
+	// TODO - should post '..' and '.' ?
+	for f.dirStream.HasNext() {
+		e, code := f.dirStream.Next()
+
+		if !code.Ok() {
+			f.dirStream.Close()
+			return code
+		}
+		if !out.AddDirEntry(e) {
+			f.overflow = e
+			f.hasOverflow = true
+			return code
+		}
+	}
+
+	f.dirStream.Close()
+	return fuse.OK
 }
 
 func (b *rawBridge) ReadDirPlus(input *fuse.ReadIn, out *fuse.DirEntryList) fuse.Status {
-	return fuse.ENOSYS
-}
+	n, f := b.inode(input.NodeId, input.Fh)
 
-func (b *rawBridge) ReleaseDir(input *fuse.ReleaseIn) {
-	return
+	if code := b.getStream(input, n, f); !code.Ok() {
+		return code
+	}
+
+	if f.hasOverflow {
+		// always succeeds.
+		out.AddDirEntry(f.overflow)
+		f.hasOverflow = false
+	}
+
+	for f.dirStream.HasNext() {
+		var e fuse.DirEntry
+		var code fuse.Status
+		if f.hasOverflow {
+			e = f.overflow
+			f.hasOverflow = false
+		} else {
+			e, code = f.dirStream.Next()
+		}
+
+		if !code.Ok() {
+			f.dirStream.Close()
+			return code
+		}
+
+		entryOut := out.AddDirLookupEntry(e)
+		if entryOut == nil {
+			f.overflow = e
+			f.hasOverflow = true
+			return fuse.OK
+		}
+
+		child, code := n.node.Lookup(context.TODO(), e.Name, entryOut)
+		if !code.Ok() {
+			if b.options.NegativeTimeout != nil {
+				entryOut.SetEntryTimeout(*b.options.NegativeTimeout)
+			}
+		} else {
+			b.addNewChild(n, e.Name, child, nil, entryOut)
+			b.setEntryOutTimeout(entryOut)
+			if (e.Mode &^ 07777) != (child.mode &^ 07777) {
+				// XXX should go back and change the
+				// already serialized entry
```

Comment:
Can we move `out.AddDirLookupEntry` after `n.node.Lookup`? This would avoid the problem

Reply:
no, because if the entry doesn't fit, we'll have done a LOOKUP (with associated nlookup increment) without telling the kernel.